src/block.ts

```diff
@@ -0,0 +1,269 @@
+
+    if (!stringError) {
+      return errors.length === 0
+    }
+
+    return errors.join(' ')
+  }
+
+  /**
+   * Validates the entire block, throwing if invalid.
+   *
+   * @param blockChain - the blockchain that this block wants to be part of
+   */
+  async validate(blockChain: Blockchain): Promise<void> {
+    await Promise.all([
+      this.validateUncles(blockChain),
+      this.genTxTrie(),
+      this.header.validate(blockChain),
+    ])
+
+    if (!this.validateTransactionsTrie()) {
+      throw new Error('invalid transaction trie')
+    }
+
+    const txErrors = this.validateTransactions(true)
+    if (txErrors !== '') {
+      throw new Error(txErrors)
+    }
+
+    if (!this.validateUnclesHash()) {
+      throw new Error('invalid uncle hash')
+    }
+  }
+
+  /**
+   * Validates the uncle's hash
+   */
+  validateUnclesHash(): boolean {
+    const raw = rlp.encode(this.uncleHeaders.map(uh => uh.raw))
+
+    return ethUtil.keccak256(raw).toString('hex') === this.header.uncleHash.toString('hex')
+  }
+
+  /**
+   * Validates the uncles that are in the block, if any. This method throws if they are invalid.
+   *
+   * @param blockChain - the blockchain that this block wants to be part of
+   */
+  async validateUncles(blockchain: Blockchain) {
+    if (this.isGenesis()) {
+      return
+    }
+
+    if (this.uncleHeaders.length > 2) {
+      throw new Error('too many uncle headers')
+    }
+
+    const uncleHashes = this.uncleHeaders.map(header => header.hash().toString('hex'))
+
+    if (!(new Set(uncleHashes).size === uncleHashes.length)) {
+      throw new Error('duplicate uncles')
+    }
+
+    return Promise.all(this.uncleHeaders.map(async uh => this._validateUncleHeader(uh, blockchain)))
+  }
+
+  /**
+   * Returns the block in JSON format
+   *
+   * @see {@link https://github.com/ethereumjs/ethereumjs-util/blob/master/docs/index.md#defineproperties|ethereumjs-util}
+   */
+  toJSON(labeled: boolean = false) {
+    if (labeled) {
+      return {
+        header: this.header.toJSON(true),
+        transactions: this.transactions.map(tx => tx.toJSON(true)),
+        uncleHeaders: this.uncleHeaders.forEach(uh => uh.toJSON(true)),
+      }
+    } else {
+      return ethUtil.baToJSON(this.raw)
+    }
+  }
+
+  private async _putTxInTrie(txIndex: number, tx: Transaction) {
+    await new Promise((resolve, reject) => {
+      this.txTrie.put(rlp.encode(txIndex), tx.serialize(), (err: any) => {
+        if (err) {
+          reject(err)
+        } else {
+          resolve()
+        }
+      })
+    })
+  }
+
+  private _validateUncleHeader(uncleHeader: BlockHeader, blockchain: Blockchain) {
+    // TODO: Validate that the uncle header hasn't been included in the blockchain yet.
+    // This is not possible in ethereumjs-blockchain since this PR was merged:
+    // https://github.com/ethereumjs/ethereumjs-blockchain/pull/47
+
+    const height = new BN(this.header.number)
+    return uncleHeader.validate(blockchain, height)
+  }
```

Comment:
I am just not getting the whole picture why this "uncle already included" could be removed, could you give a really quick reasoning?

Reply:
It shouldn't be removed, but there's no way to validate it using `ethereumjs-blockchain`. The previous code was just calling a function of the blockchain that was a no-op. 

This really surprised me, so I spent some time trying to figure out how it happened, but couldn't found the reason. All I know is that that functionality was removed in this PR: https://github.com/ethereumjs/ethereumjs-blockchain/pull/47/files#diff-168726dbe96b3ce427e7fedce31bb0bcR481

Reply:
//cc @vpulim